### PR TITLE
Remove StringBuffer and use ValueStringBuilder

### PR DESCRIPTION
**Pull request by JeremyKuhne**
cc: @danmosemsft

### CODE REVIEW

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 1]
```diff
@@ -24,63 +25,73 @@ internal class PathHelper
         /// <returns>Normalized path</returns>
         internal static string Normalize(string path)
         {
-            // Get the full path
-            StringBuffer fullPath = new StringBuffer(path.Length);
+            if (path.Length < PathInternal.MaxShortPath)
+            {
+                Span<char> initialBuffer = stackalloc char[PathInternal.MaxShortPath];
+                return Normalize(path, new ValueStringBuilder(initialBuffer));
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: I would be cleaner to pass the ValueStringBuilder by ref, and also Dispose it here.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 2]
```diff
@@ -24,63 +25,73 @@ internal class PathHelper
...
+                    if (builder.AsSpan().Equals(path.AsReadOnlySpan()))
                     {
                         // If we have the exact same string we were passed in, don't bother to allocate another string from the StringBuffer.
                         return path;
                     }
-                    return fullPath.ToString();
+                    return builder.ToString();
                 }
             }
             finally
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: I do not think we need to worry about cleanup on exceptional paths.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 3]
```diff
@@ -24,63 +25,73 @@ internal class PathHelper
...
         }
 
-        private static unsafe void GetFullPathName(string path, ref StringBuffer fullPath)
+        private static void GetFullPathName(string path, ref ValueStringBuilder builder)
         {
             // If the string starts with an extended prefix we would need to remove it from the path before we call GetFullPathName as
             // it doesn't root extended paths correctly. We don't currently resolve extended paths, so we'll just assert here.
             Debug.Assert(PathInternal.IsPartiallyQualified(path) || !PathInternal.IsExtended(path));
+            builder.EnsureCapacity(path.Length);
+            builder.Append('\0');
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Why are we appending the zero? It does not seem necessary.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 4]
```diff
@@ -24,63 +25,73 @@ internal class PathHelper
         /// <returns>Normalized path</returns>
         internal static string Normalize(string path)
         {
-            // Get the full path
-            StringBuffer fullPath = new StringBuffer(path.Length);
+            if (path.Length < PathInternal.MaxShortPath)
+            {
+                Span<char> initialBuffer = stackalloc char[PathInternal.MaxShortPath];
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: I didn't think it would cost much- we're not zeroing out anymore, right? Presumably the extra space on the stack isn't important. Any rule of thumb as to what is worth avoiding?

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: So let the buffer go free and remove the finally? I'm assuming you're referring to better perf without the try block?

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 3 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: I could set the length to 1. We have an assert that doesn't allow you index beyond the length. Thoughts?

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 5]
```diff
@@ -24,63 +25,73 @@ internal class PathHelper
...
+            if (result == 0)
+            {
+                // Failure, get the error and throw
+                int errorCode = Marshal.GetLastWin32Error();
+                if (errorCode == 0)
+                    errorCode = Interop.Errors.ERROR_BAD_PATHNAME;
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, path);
             }
+
+            builder.Length = checked((int)result);
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Nit: This does not need to be checked.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 6]
```diff
@@ -89,37 +100,34 @@ private static int GetInputBuffer(ref StringBuffer content, bool isDosUnc, ref S
...
+                buffer.Append(PathInternal.UncExtendedPathPrefix);
 
-                // Copy the source buffer over after the existing UNC prefix
-                content.CopyTo(
-                    bufferIndex: PathInternal.UncPrefixLength,
-                    destination: ref buffer,
-                    destinationIndex: PathInternal.UncExtendedPrefixLength,
-                    count: content.Length - PathInternal.UncPrefixLength);
+                // Copy Server\Share\... over to the buffer
+                buffer.Append(buffer.AsSpan().Slice(PathInternal.UncPrefixLength));
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: This does not look right. The buffer is only going to have `UncExtendedPathPrefix` in it at this point, and so the Slice will be alway empty.

Comment (jkotas) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Should it be `content`?

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Because per spec we can't get an overflow value? (Should never ask for more than UNICODE_STRING can hold). I was thinking technicalities here as the API returns uint by definition.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 2 above)
Comment (jkotas) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Right.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 6 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Yes, good catch. Apparently I'm missing a test case here (UNC with short name).

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 3 above)
Comment (jkotas) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: You can use `Marshal.GetReference(...)` to get the pointer instead of taking address of the element zero. E.g. Add `ref char GetPinnableReference() => Marshal.GetReference(_chars)` method to the ValueStringBuilder. Use that instead of taking address of element zero. It will be also a bit smaller code because of it will avoid the useless boundscheck against zero.

Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Will do that, thanks.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 5 above)
Comment (jkotas) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: The while loop above will exit only if the total size is less than the capacity. So it has to fit into int.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Smaller code with fewer special cases is better.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: :) thanks!

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 7]
```diff
@@ -131,10 +115,10 @@ private static string TryExpandShortFileName(ref StringBuffer outputBuffer, stri
             //
             // Note that we will never get \??\ here as GetFullPathName() does not recognize \??\ and will return it as C:\??\ (or whatever the current drive is).
 
-            int rootLength = PathInternal.GetRootLength(outputBuffer.AsSpan());
-            bool isDevice = PathInternal.IsDevice(outputBuffer.AsSpan());
+            int rootLength = PathInternal.GetRootLength(outputBuilder.AsSpan());
+            bool isDevice = PathInternal.IsDevice(outputBuilder.AsSpan());
 
-            StringBuffer inputBuffer = new StringBuffer(0);
+            ValueStringBuilder inputBuilder = new ValueStringBuilder();
```
Comment (stephentoub) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Is it not worth starting with stack space?

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 8]
```diff
@@ -194,30 +181,30 @@ private static string TryExpandShortFileName(ref StringBuffer outputBuffer, stri
...
                         }
                     }
-                    else if (result > outputBuffer.Capacity)
+                    else if (result > outputBuilder.Capacity)
                     {
                         // Not enough space. The result count for this API does not include the null terminator.
-                        outputBuffer.EnsureCapacity(checked((int)result));
-                        result = Interop.Kernel32.GetLongPathNameW(inputBuffer.UnderlyingArray, outputBuffer.UnderlyingArray, (uint)outputBuffer.Capacity);
+                        outputBuilder.EnsureCapacity(checked((int)result));
+                        result = Interop.Kernel32.GetLongPathNameW(ref inputBuilder[0], ref outputBuilder[0], (uint)outputBuilder.Capacity);
```
Comment (stephentoub) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Can this be GetPinnableReference?

File: src/mscorlib/shared/System/Text/ValueStringBuilder.cs [hunk 9]
```diff
@@ -21,7 +22,25 @@ public ValueStringBuilder(Span<char> initialBuffer)
...
 
-        public int Length => _pos;
+        public int Length
+        {
+            get => _pos;
+            set
+            {
+                EnsureCapacity(value);
+                _pos = value;
+            }
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Text/ValueStringBuilder.cs]: What does it mean to set the length like this? Couldn't this be filled with garbage?

File: src/mscorlib/shared/System/Text/ValueStringBuilder.cs [hunk 10]
```diff
@@ -141,6 +162,18 @@ public unsafe void Append(char* value, int length)
...
 
+        public unsafe void Append(ReadOnlySpan<char> value)
+        {
+            int pos = _pos;
+            if (pos > _chars.Length - value.Length)
+            {
+                Grow(value.Length);
+            }
+
+            value.CopyTo(_chars.Slice(_pos, value.Length));
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Text/ValueStringBuilder.cs]: It shouldn't be necessary to pass value.Length.

File: src/mscorlib/shared/System/Text/ValueStringBuilder.cs [hunk 11]
```diff
@@ -164,7 +197,7 @@ private void GrowAndAppend(char c)
         [MethodImpl(MethodImplOptions.NoInlining)]
         private void Grow(int requiredAdditionalCapacity)
         {
-            Debug.Assert(requiredAdditionalCapacity > _chars.Length - _pos);
+            Debug.Assert(requiredAdditionalCapacity > 0);
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Text/ValueStringBuilder.cs]: Why did you change this?

File: src/mscorlib/shared/System/Text/ValueStringBuilder.cs (hunk 9 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Text/ValueStringBuilder.cs]: I would drop the `EnsureCapacity` here - require the space to be preallocated and filled in by other means.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 7 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: I didn't want to add another .5k to the stack for the corner case. I can if you think I should.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 8 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Yes, I missed this, thanks!

File: src/mscorlib/shared/System/Text/ValueStringBuilder.cs (hunk 9 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/Text/ValueStringBuilder.cs]: Will do.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 7 above)
Comment (jkotas) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Maybe worth a comment.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 12]
```diff
@@ -233,26 +221,26 @@ private static string TryExpandShortFileName(ref StringBuffer outputBuffer, stri
...
             }
             finally
             {
-                inputBuffer.Free();
+                inputBuilder.Dispose();
             }
         }
 
         // Helper method to workaround lack of operator ? support for ref values
-        private static ref StringBuffer Choose(bool condition, ref StringBuffer s1, ref StringBuffer s2)
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: You should not need this method anymore. The current C# compiler used for coreclr/corefx should have support for ?: for ref values. I think it should be something like:
```
ref ValueStringBuilder bufferToUse = ref (success ? ref outputBuilder : ref inputBuilder);
```

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 13]
```diff
@@ -233,26 +221,26 @@ private static string TryExpandShortFileName(ref StringBuffer outputBuffer, stri
...
                 }
                 else
                 {
-                    returnValue = bufferToUse.Substring(rootDifference, newLength);
+                    returnValue = new string(bufferToUse.AsSpan().Slice(rootDifference, newLength));
                 }
 
                 return returnValue;
             }
             finally
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Nit: Remove try/finally ?

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 14]
```diff
@@ -89,173 +76,152 @@ private static int GetInputBuffer(ref StringBuffer content, bool isDosUnc, ref S
...
             }
-        }
 
-        // Helper method to workaround lack of operator ? support for ref values
-        private static ref StringBuffer Choose(bool condition, ref StringBuffer s1, ref StringBuffer s2)
-        {
-            if (condition) return ref s1;
-            else return ref s2;
+            // If we were able to expand the path, use it, otherwise use the original full path result
+            outputBuilder = success ? ref outputBuilder : ref inputBuilder;
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: This will *overwrite* the outputBuilder with inputBuilder. The result is going to be that we are not going to release the buffer from the original outputBuilder if there is one.

I think you still need `ref StringBuffer bufferToUse` local here like before.

Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: D'oh.  If I take out the ref on the method sig will that solve it?

Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: On second thought, that might not be a good idea. I could see that getting reintroduced.

Comment (jkotas) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: I do not think it would solve it because of you are mutating the outputBuilder here.

Also, I do not think it is a generally good idea to play tricks when you do copy vs. when you do not make a copy of "no copy" types. It is best to always pass the "no copy" types by ref. I wish there was analyzer to check for copies of no-copy types.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 15]
```diff
@@ -89,173 +76,152 @@ private static int GetInputBuffer(ref StringBuffer content, bool isDosUnc, ref S
...
             }
-        }
 
-        // Helper method to workaround lack of operator ? support for ref values
-        private static ref StringBuffer Choose(bool condition, ref StringBuffer s1, ref StringBuffer s2)
-        {
-            if (condition) return ref s1;
-            else return ref s2;
+            // If we were able to expand the path, use it, otherwise use the original full path result
+            ValueStringBuilder builderToUse = success ? outputBuilder : inputBuilder;
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: This is still creating a copy. It is non-inefficient and dangerous (there is no actual problem right now because of nothing seems to mutating the builderToUse).

Could you please change this to `ref ValueStringBuilder builderToUse = ...` ?

Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Sure. I clearly haven't gotten my sea-legs with the new ref syntax. :/
